shell/platform/android/io/flutter/embedding/FlutterRenderer.java

```diff
@@ -0,0 +1,284 @@
+  }
+
+  public void attachToRenderSurface(@NonNull RenderSurface renderSurface) {
+    // TODO(mattcarroll): what is our desired behavior when attaching to an already attached renderer?
+    if (this.renderSurface != null) {
+      detachFromRenderSurface();
+    }
+
+    this.renderSurface = renderSurface;
+  }
+
+  public void detachFromRenderSurface() {
+    // TODO(mattcarroll): do we care if we're asked to detach without first being attached?
+    if (this.renderSurface != null) {
+      this.renderSurface = null;
+    }
+  }
+
+  public void addOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedListener listener) {
+    firstFrameListeners.add(listener);
+  }
+
+  public void removeOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedListener listener) {
+    firstFrameListeners.remove(listener);
+  }
+
+  private void notifyFirstFrameListeners() {
+    for (OnFirstFrameRenderedListener listener : firstFrameListeners) {
+      listener.onFirstFrameRendered();
+    }
+  }
+
+  //------ START TextureRegistry IMPLEMENTATION -----
+  // TODO(mattcarroll): this method probably shouldn't be public. It's part of an obscure relationship
+  //   with PlatformViewsController. Re-evaluate that relationship and see if we can get rid of
+  //   PlatformViewsController.
+  // TODO(mattcarroll): detachFromGLContext requires API 16. what should we do for earlier APIs?
+  @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+  @Override
+  public SurfaceTextureEntry createSurfaceTexture() {
+    final SurfaceTexture surfaceTexture = new SurfaceTexture(0);
+    surfaceTexture.detachFromGLContext();
+    final SurfaceTextureRegistryEntry entry = new SurfaceTextureRegistryEntry(
+        nextTextureId.getAndIncrement(),
+        surfaceTexture
+    );
+    registerTexture(entry.id(), surfaceTexture);
+    return entry;
+  }
+
+  final class SurfaceTextureRegistryEntry implements TextureRegistry.SurfaceTextureEntry {
+    private final long id;
+    private final SurfaceTexture surfaceTexture;
+    private boolean released;
+
+    SurfaceTextureRegistryEntry(long id, SurfaceTexture surfaceTexture) {
+      this.id = id;
+      this.surfaceTexture = surfaceTexture;
+      this.surfaceTexture.setOnFrameAvailableListener(new SurfaceTexture.OnFrameAvailableListener() {
+        @Override
+        public void onFrameAvailable(SurfaceTexture texture) {
+          markTextureFrameAvailable(SurfaceTextureRegistryEntry.this.id);
+        }
+      });
+    }
+
+    @Override
+    public SurfaceTexture surfaceTexture() {
+      return surfaceTexture;
+    }
+
+    @Override
+    public long id() {
+      return id;
+    }
+
+    @Override
+    public void release() {
+      if (released) {
+        return;
+      }
+      unregisterTexture(id);
+      surfaceTexture.release();
+      released = true;
+    }
+  }
+  //------ END TextureRegistry IMPLEMENTATION ----
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
```

Comment:
TMK these are methods called by Android system when the texture lifecycle changes (occluded/needs redrawing etc).

Reply:
@mehmetf any idea if there are docs about that lifecycle?  I'd like to comment here with the technical specifics of what it means for a surface to be created, changed, or destroyed...